### DISCUSSION TITLE
[question] what do you suggest for windows 7 users ?

### DISCUSSION THREAD

**Discussion by John0King** [dotnet/maui#3051] [Q&A] [answered]
### Description

what do you suggest for windows 7 users ?  and also  @WinUI-team .

Will you suggest me to use flutter /  electron  and goes away from .net platform ? or you suggest me stay in old place and can't get in to the newer .net community (separate .net community  to  two smaller community  named `old` and `new`   ) ?

and I feel  whole windows develop happened in MS  is a just like this image 

![黑瞎子掰棒子](https://user-images.githubusercontent.com/13639146/137657699-fceb6093-f1a8-4c18-a1cf-e4483d0aa821.jpg)

and you get more 1 and drop more 1 too.

**Comment by saint4eva** [comment]
Windows 7 is going out of support.  Does flutter support it? Kindly encourage your users to upgrade to probably Windows 11.

**Comment by John0King** [comment]
flutter use graphics engine (`skia`) to drawing every  widgets ,  there no reason they cut windows support，   and windows 7 is supported until  2023/1/1,  and even after that time  ,  the support for those users still depends on  how many user use it instead of MS said they will not support win 7 or not .  
and think about that if where are 30% users still use windows 7 ,  will  Youtube/steam/twitch/facebook/EA/unity3d  drop 30% users ?

and according to the data  [here](https://gs.statcounter.com/windows-version-market-share/desktop/worldwide/#monthly-202009-202109), there are still nearly 15% windows 7 usres

**Comment by saint4eva** [comment]
Then use flutter - it would better fit your need.

**Reply by AlexByte** [reply]
there is some other language

**Comment by John0King** [comment]
@saint4eva   what's wrong with Win7 ?
There is a `silk.net`   why wont build a UI on that ?  #2576 
and look at vs code and electron ,  they use HTML to draw 的 UI , and it has notification, custom windows ,  many windows api ( I mean many  api that win7 do not have that can be mock with custom UI/logic )

**Reply by saint4eva** [reply]
@John0King there is NOTHING wrong with Windows 7 - in fact Win 7  is one of the greatest operating systems (OSes) in existence.
But, .NET Maui team has priorities and of which supporting Windows 7 is not high on the measurement scale. 

Mind you, Maui just targets the respective OSes using their recommended native application development platforms. E.g. WinUI/ WinAppSDK for Windows, Catalyst for MacOS etc. So, if WinUI does not have proper support for Win 7, then it would be hard for Maui to target it - not that it is not doable. 

But if you really need support for Win 7, UNO Platform will give you all the necessary tooling and support.

**Reply by John0King** [reply]
start yeas before , windows team start  create gaps between each windows version, and now they create another gap between win 10 and win 11 just because they want to increase the sales of  new computers.
and today for company use computer not like 10 yeas ago,  most of the enterprise user only use computer do run a few company software, and  though  win 7 is old but  that nothing wrong with win7 as you said.     yes , today you can give up win7  and what about tomorrow ?  win10 is the next ,  eventually everything old in ages,    the policy either  lax like android that make user self upgrade  or like windows  phone  that introduce gaps .    I think you still can use latest android studio and  android sdk  to create an app run on android 4  (but there no more phones run on a android bellow 6 right ?)      they even create some tool to make  `andoridX` adapted to old  `andorid support` library (and xamarin also do something similar)

and this is what I fighting of,   because .net **always the first** language that drop  OS support, that make  .net like a weapon to control custom by control the developers .  and MS also said .net is an choice  too, does means : ".net is just option for build app "  while other language means "<the lang/platform> is what you need to build the apps"  ,    and eventually make us those .net devs like idiot when you just meet this problem.  and Ask you self :  wouldn't you want to build app that everyone can run your work result（the app）?  or you just want drop someone that use old OS.

I will be convinced if there are really problem to support the old OS, but is there any problems?   

like I ask before, there are just tested solution there like electron/uniapp/flutter/nativescript,  will you choose MAUI if you really want to service your custom  ?
MAUI:  `new/ smaller community/ limited windows OS support (drop 15% market) / no web app support/ no mini apps support`
flutter: `bigger community/be tested by other apps/ mobile  and web apps / desktop preview  / no windows OS version limit`
electron: `much bigger community/ Vue-react-Angular framework/ w3c standard group /  no windows OS version limit/ cross platfrom everywhere`
....

**Reply by CommonLoon102** [reply]
Flutter: it isn't really good for web apps, you press Ctrl+F and can't find anything because even the text is just a PICTURE.
Electron: it is based on Chromium and Google is only planning to support Chrome (and therefore probably Chromium as well) on Windows 7 until 2023 January 15. (https://cloud.google.com/blog/products/chrome-enterprise/extending-chrome-on-windows-7-to-support-enterprise-customers).

There isn't really a good solution which fits everything, it really depends on what you want to build. If the target is Windows 7-11 only, then you are fine with .NET 6 (WPF / WinForms) for an another 2 years.

**Reply by John0King** [reply]
@CommonLoon102   that wan't the truth,   do you know with same codebase,   Tencent/Alibaba build their apps using electron even run on windows xp and also with win10/win11 native notification ,  there is typescript/babel and profill scripts that make old browser also run you latest code.

now , even Microsoft love those stack (skype/teams/vscode they a few of them maybe using react native, but the process should be the same with V8 engine)

but can you do the same with WPF/winform?  after 2 yeas , you had to choice again and again  and again.  so I ask myself many times: "since .net isn't a trustable platform   why you still using it instead of go to thing like electronic  ?"    
I think maybe C# is very a good language .   but my faith won't live forever if .net's stack continue  introduce the gaps that can not be filled.

**Reply by CommonLoon102** [reply]
Actually MS doesn't like electron.js and tries to move away from it. MS Teams won't be based on electron in a few years, it wastes to much memory.
https://techcommunity.microsoft.com/t5/microsoft-teams/teams-2-0-moves-away-from-electron-to-embrace-edge-webview2/td-p/2484565

If the new electron versions will use the new chromium versions, then it won't work anymore on Windows 7 after 2023. You can still use the old electron as long as you want, like you can use VB.NET and WindowsForms.NET on .NET Framework 2.0 on Windows 11 if you would like to.

**Reply by John0King** [reply]
the different is that you do not need to install other "runtime" with electron with both new and old version,  
the the important is that no code change between old and new ,  the only change is a configuration that effect typescript/babel translate .
teams move away from a V8 + HTML render  to another V8+ HTML render , there is nothing change.

**Reply by CommonLoon102** [reply]
There are new features in newer versions of electron.js, if you want to use something new, for example new native calls to Windows 11 (new kind of notifications/whatever). So you would need to maintain two versions of your app, one is bundled with the newest electron.js, the other is bundled with an old one which still runs on Windows XP.

Regarding MS Teams, I guess they want to offload some parts of it to native handlers, like video chats, opening files and documents, etc. This way not the browser would handle most of the things and therefore it might be less memory intensive. But who knows.

**Reply by CommonLoon102** [reply]
No more Electron for Windows 7 either:
https://www.electronjs.org/blog/windows-7-to-8-1-deprecation-notice

**Reply by John0King** [reply]
there nothing effect when it drop support windows 7,  because most of features that the old electron/browser don't have can be simply plug a babel translater or a polyfill,  you can still write the newest javascript/typescript syntax  and target both for windows 7 and newer.   and that is not the same for .net . you have to choose :   stay behind or move forward.    stay behind means you can target both old and new platform but without the new platfrom feature; move forward means you drop many users/customs.

**Reply by CommonLoon102** [reply]
Up until TLS 1.3 becomes deprecated and all the servers stop using it. But that might be decades. Or months. Who knows.
https://en.wikipedia.org/wiki/Transport_Layer_Security#Cipher

https://www.chromium.org/Home/tls13/

**Reply by CommonLoon102** [reply]
Actually, Windows 7 doesn't even support TLS 1.3, so when TLS 1.2 gets obsoleted and servers start using TLS 1.3, your electron apps on Windows 7 are going to stop. Except if you show local HTML files only.

https://stackoverflow.com/questions/62677922/is-there-a-way-for-windows-7-to-support-tls-1-3-net-4-8

**Comment by GeraudFabien** [comment, marked as answer]
https://github.com/jsuarezruiz/maui-linux

**Reply by CommonLoon102** [reply]
MAUI-Linux is not even in beta.

@John0King Have you checked out the Uno Platform?
https://platform.uno/

It has Win 7 support, and also web, mobile, win10, linux, mac.

It is even better than Flutter, because you can do Ctrl+F in the browser and you can search for text.
You can also set the text selectable: https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.textblock?view=winrt-22000
Copying also works when you select text.
Check out TextBlock in the Uno Gallery: https://gallery.platform.uno/

The only downside is the memory usage (at least on web for sure). If I open gallery.platform.uno then that tab consumes 2.2 GB in Chrome. That might be a problem if you want to open the site on mobile. But for mobile there should be an app version deployed.
Even a simple calculator consumes 500 MB: https://calculator.platform.uno/

**Comment by CommonLoon102** [comment]
What is wrong with .NET 4.8 and WindowsForms.NET for Win 7? It still runs on Win 11 too.

**Reply by John0King** [reply]
there are nothing "wrong" with them,   the "wrong" come with ages and against developers "move forward" ,  just like nothing wrong with `python1` and `python2`  or things like `assembly language` .

As you know ,  .Net 4.8 has no support .net standard 2.1, that will effect library usage,  and it will be two isolate  community if you developed library can not be consume  by  the new .net , and other group of developers from  new .net developed library can not be consume by you.  I called those behaviors "drop behind" / “abandoned”.

Windows team create multiple gaps from windows 7  to  windows 11,  but there no gaps in .net  from win7 to win 11. 

------------
I find your are suggesting https://github.com/jsuarezruiz/maui-linux  to also support windows using GTK3, and this is the same thing that I asked here (with a few different goal).

-  .net team shouldn't follow windows team's “gaps” when where is a solution can be used to fill it (like typescript does for ES3/ES5 from ES2020)
-  compare to GTK that using native library, there is a `silk.net` which I think it is the graphic engine in strid3d( Xenko ) and is now maintained by .net foundation  , but GTK is also greate choice .  (and   GUI app is almost a Canvas + a few native api, like electron/flutter/uniapp/NativeScript does)

**Reply by CommonLoon102** [reply]
Actually .NET 6 runs just fine on Windows 7, so you could create a WPF (or WinForms) app on .NET 6 and then you will have access to the latest libraries, etc.

It's only the MAUI thing which wouldn't run, but what's good in it? It is mostly for mobile development (Android, iOS) plus it runs on Win10 so they don't have to emulate Android apps (what Win 11 can already do anyways).

If they really wanted to do a real cross-platform GUI ever, they would have already done it 10-20 years ago.

**Reply by John0King** [reply]
> then you will have access to the latest libraries

Windows App SDK is also have gaps I think. it doesn't support windows 7 even things like camera or blue-tooth/usb  (while those are became standard api in browse)

**Reply by John0King** [reply]
> If they really wanted to do a real cross-platform GUI ever, they would have already done it 10-20 years ago

totally agreed

**Reply by CommonLoon102** [reply]
Have you checked out platform.uno?

**Reply by John0King** [reply]
tried on the web with the Xaml player , it's cool.

**Comment by Xyncgas** [comment]
For any platform, understand the display. For a 2D display you can write algorithms in any language you'd like, then proceed to drawing layout on the screen using whatever way you can. For Windows 7 .Net Framework worked in history, so has browsers, so has QT. Now you can create a layer that draw any rectangle you would like on the 2D display with the algorithms you've created on the UI framework (tools that utilize the operating system's api for UI) of your choice. Meanwhile absolute positioning was the only thing you needed.

Step 1. Write your algorithm down in a flow chart (E.g. how to crop a rectangle with another rectangle, how to automatically fill a rectangle with small rectangle of given height & width, how to relocate a corner of a rectangle to a new place...)

Step 2. Pick your favorite programming language and implement the algorithms (Fun)

Step 3. Pick the UI Framework you like (.NET has an ecosystem), find out how to draw a rectangle using it using absolute layout, write glue codes which means finding out how to convert the result of your algorithm to draw the rectangle on screen

Step 4. Build your application

Step 5. Refactoring or changing your algorithm, or start over using another UI framework that you like better but you can keep the algorithms you wrote just rewrite the glue codes

**Reply by CommonLoon102** [reply]
Writing a proper text renderer with support of several fonts, ClearType, screen readers for the visually impaired, plus solving the issue of text flow and alignment itself would be a work for years. Good luck.

**Reply by CommonLoon102** [reply]
If you don't want to reinvent the wheel, use Skia: https://github.com/google/skia
Still, you don't have OS integration, then just use AvaloniaUI or platform.uno.
https://github.com/AvaloniaUI/Avalonia
https://platform.uno/
